README.md
####Extended Safezones with Vehicle Protection for A3-EPOCH
<br/>
- Title: `SPK_ESVP`
- Author: `Sp4rkY` [Link](https://github.com/SPKcoding)
- Description: `Extended Safezones with Vehicle Protection for Arma 3 Epoch`
- Version: `0.916`
- Required: `Arma 3 1.66+ / Epoch 0.4+`
- Credits:<br/>`IT07 (Thx for many brain)` [Link](https://github.com/IT07)<br/>`|V.I.P.| CH!LL3R (Thx for thoughts & testing!)`<br/>

___
<br/>
######FEATURES:<br/>
- Fully customizable<br/>
- Works with all maps<br/>
- Works for players and vehicles<br/>
- No more `"ProtectionZone_Invisible_F"` is needed (usually found in map config in a3_epoch_server_settings.pbo)<br/>
- Fired bullets will automatically be deleted in safezones<br/>
- (optional) Teleport players out of safezones after restart<br/>
- (optional) Teleport players away from the parking place (if used) after restart<br/>
- Protection from idiots trying to drive over other players (including traders)<br/>
- Check vehicle ownership of bought vehicles<br/>
- (optional) Teleport vehicles out of safezones at server restart to:<br/>
	* (optional) a given distance from the respective safezone<br/>
	* (optional) a static parking place (you can use your own map addition and define as many parking lots you want)<br/>
* (optional) Restricted vehicle´s access in safezones (only the owner(s) can enter the vehicle)<br/>
* (optional) Restricted vehicle´s gear-access in safezones (only the owner(s) can access the vehicle´s inventory)<br/>
* (optional) Protection for lifted vehicle´s<br/>
	* If using this feature, there will be also a check for the owner, so stealing is not possible<br/>
* (optional) Unlock teleported vehicles<br/>
* 2 ways of info messages (depends on if vehicle protection is used or not)<br/>
* (optional) Vehicle Access Menu:<br/>
	* "Refresh" button to update the player-list while menu is open<br/>
	* "Registered" listing to see which players are added to your vehicle<br/>
	* "Clear" button to remove the added players<br/>
* (optional) restrict access to driver seat for primary vehicle owner<br/>
* slingload check for bought vehicles (you can not steal cars or ships)<br/>
* (optional) info messages for prohibited lifting<br/>
* (optional) Prevent spawning of antagonists (UAV, Sappers, Snakes, Cultists) in safezones<br/>
* (optional) Allow/disallow to chop/sledge/chainsaw anything in safezones<br/>
* (optional) Built-in map marker feature with most needed options<br/>
* If using the serverside "vehicle teleporting WITHOUT Parking Place" - function:<br/>
	* the vehicles get teleported to a random position within the given radius out of the safezones<br/>
* If using the serverside "vehicle teleporting WITH Parking Place" - function, it will works like this:<br/>
	* server restarts, vehicles in safezones will be teleported to a parking place<br/>
	* vehicles which are present at the parking place are getting teleported to a random location on the map, new vehicles (in safezones) are ported to the parking place<br/>
* If the respective safezone contains water, ships are treated like this:<br/>
	* if ships are located at water surface, they gets teleported close to the safezone´s shore at given radius<br/>
	* if ships are located at terrain surface because of lifting or simliar, they gets ported random at the maps shore<br/>
* If not using "vehicle teleporting" - function, remaining vehicles in safezones are indestructible after restart<br/>
* Debug option for better determine errors (client & server)<br/>
<br/>

___
<br/>

######INSTALL INSTRUCTIONS:<br/>
`Assumed you know how to use a Pbo-Manager and Notepad++ or similar tools. Otherwise click the following link:`<br/>
[Arma: Community Tools](https://community.bistudio.com/wiki/ArmA:_Community_Tools)
<br/>

`Once you have downloaded and extracted the files, start like this:`

_SERVER:_
<br/>
* Go to the downloaded folder `@epochhive\addons\epoch_SPK_ESVP\`
* Use your favourite text/C++ editor, open the `settings.h` and customize it to your wishes (every option should be self-explaining by looking at their comments)
* Save the file
* Open `epoch_SPK_ESVP.pbo` with your PBO-Manager & delete all files in it
* Copy & Paste all files of the `epoch_SPK_ESVP` folder into the `epoch_SPK_ESVP.pbo`
* Upload this PBO to your servers `@epochhive\addons\` folder
<br/>
* New way of enabling SPKcode scripts:
* Upload the `epoch_SPKcode_config.pbo` also to your servers `@epochhive\addons\` folder
* At the moment there is no need to cange anything in this PBO, just upload it and you are good to go. (will be updated as soon as more scripts are available)

<br/>

_CLIENT:_
<br/>
`If you have any problems with following these instructions, take a look at the "mpmissions\epoch.Mapname" folder of this package. Every file which needs to get edited you can find there`<br/>
<br/>
* Download and unPbo your `epoch.Mapname.pbo` ("Mapname" depends on which map you use (example: epoch.Altis.pbo))
* Open your `epoch.Mapname\description.ext` and place the following at the bottom:
```C++
#include "SPKcode\config.cpp"
```
* Open your `epoch.Mapname\epoch_config\Configs\CfgFunctions.hpp` and place the following `before` the last closing bracket:
```C++
#include "SPK_CfgFunctions.hpp"
```
* Copy the file `epoch.Mapname\epoch_config\Configs\SPK_CfgFunctions.hpp` to the same directory in your missionfile
* Copy the file `epoch.Mapname\epoch_config\Configs\CfgRemoteExec.hpp` to the same directory in your missionfile (overwrite the existing one)
	* it´s prepared for Epoch 0.4 (if you already have custom changes in it, just merge it)
* Copy the file `epoch.Mapname\epoch_config\Configs\SPK_CfgRemoteExec.hpp` to the same directory in your missionfile
* If you already have a `class CfgNotifications` anywhere in your missionfile, just open the file `epoch.Mapname\SPKcode\config.cpp` and comment out the line `#include "notes.h"` with `//` 
	* after that merge the content of `epoch.Mapname\SPKcode\notes.h` with your existing `class CfgNotifications`
* If you wish to prohibit using of melee weapons while in safezones, copy the file `epoch.Mapname\epoch_code\compile\functions\EPOCH_fnc_playerFired.sqf` to the same directory in your missionfile (overwrite the existing one)
	* it´s prepared for Epoch 0.4 (if you already have custom changes in it, just merge it)
* Copy the folder `epoch.Mapname\SPKcode` from this package into the root of your `epoch.Mapname`
* Open the file `epoch.Mapname\SPKcode\ESVP\config.cpp` in your mission and customize the client functions to your wishes. Read the self-explaining comments
* Once you are ready with customization, repack your missionfile `epoch.Mapname` and upload it back to your server
<br/>

<br/>

_INFISTAR:_
<br/>
- You have to set the following options like this:
```C++
/*  Revert allowDamage   */ _RAD = false;
/*  Revert HandleDamage  */ _RHD = false;
/*  Do not change EH_Fired at all! */ _NO_EHF = true;
```

<br/>

_BATTLEYE:_
<br/>
* Check the battleye folder in this package and add the specific filters to your ones
<br/>

___
<br/>

######ADDITIONAL INFORMATIONS:

_Standard safezones:_
<br/>
- You don´t need to use `"ProtectionZone_Invisible_F"` anymore. (normally defined in "a3_epoch_server_settings\configs\maps\map.h")
<br/>

_Parking place addition:_
<br/>
- The existing parking place addition was made for Altis, it´s located in the salt desert. It´s just an example, so you can use it or create your own. :)
	* In case you want to create your own parking place addition:
		* for every single parking lot an empty helipad is needed `Land_HelipadEmpty_F`
		* you can create as many empty_helipads as you want, you only have to insert each coordinate (x,y,z) into the serverside configfile (settings.h)
<br/>

_Prevent antagonists in safezones:_
<br/>
- If you notice a bug of the trader missions, related to the UAV´s, simply remove `"I_UAV_01_F"` from the serverside configfile (settings.h)
<br/>

___
<br/>

######VERY IMPORTANT:
- The safezone coordinates in the serverside settings.h `HAVE TO` be exactly the same as in the clientside configfile, except the name of each zone !!!
<br/>
_Example:_
<br/>
* client: `[NAME, COORDS, RADIUS]`
* server: `[COORDS, RADIUS]`
<br/>
- The serverside addon was optional, now you have to use it!
<br/>
- Added a second serverside PBO just for enable/disable SPKcode scripts easily.
<br/>

___
<br/>

######W.I.P:
- support for IgiLoad or other Towing scripts
- integration of Epoch Group System
<br/>

___
<br/>

######BUGS & ERRORS:
- Teleporting players produce a serverside rpt entry if using EpochAH (but the function is working well)
- No other known bugs present so far but if you notice some, please report them to this Git or to the respective forum thread. I will try to fix them asap. Thx
<br/>

___
<br/>


changelog.md
####Changelog
<br/>

`v0.916`<br/>
	* [fixed] added missing battleye filters for the map markers<br/>
<br/>
`v0.915`<br/>
	* [added] feature to set map markers directly from this script (no need to define them anywhere else now)<br/>
<br/>
`v0.894`<br/>
	* [fixed] missing activation of ppEffects while gui is open<br/>
<br/>
`v0.893`<br/>
	* [fixed] projectiles of armed vehicles are still getting deleted after leaving safezones<br/>
	* [updated] BE filters (scripts.txt)<br/>
<br/>
`v0.891`<br/>
	* [fixed] small syntax error in the FSM<br/>
<br/>
`v0.89`<br/>
	* [added] option to allow/disallow to chop/sledge/chainsaw anything in safezones<br/>
	* [added] option to teleport players out of safezones after restart<br/>
	* [added] option to teleport players away from the parking place after restart<br/>
	* [added] check vehicle ownership of bought vehicles<br/>
	* [added] two ways of info messages (depends on if vehicle protection is used or not)<br/>
	* [added] Vehicle Access Menu:<br/>
	* 		- "Refresh" button to update the player-list while menu is open<br/>
	* 		- "Registered" listing to see which players are added to your vehicle<br/>
	* 		- "Clear" button to remove the added players<br/>
	* [added] option to restrict access to driver seat for primary vehicle owner<br/>
	* [added] slingload check for bought vehicles (you can not steal cars or ships)<br/>
	* [added] info messages for prohibited lifting (optional)<br/>
	* [added] classnames of Ryan´s Zombies to the array for deleting antagonists in safezones<br/>
	* [added] debug option for better determine errors (client & server)<br/>
	* [fixed] vehicle lifted check is also running if _useVehProt is false<br/>
	* [fixed] vehicle protection doesn´t works for ships<br/>
	* [fixed] ability to kill players with sledgehammer in safezones<br/>
	* [fixed] bug with godmode for choppers in safezone<br/>
	* [fixed] serverside typos<br/>
	* [fixed] if not using vehicle TP, ships still have godmode<br/>
	* [fixed] major FSM overhaul<br/>
	* [fixed] major code optimization<br/>
	* [changed] client settings is now a real missionConfigFile<br/>
	* [changed] CfgFunctions now via #include for easy setup<br/>
	* [changed] serverside addon was optional, now it´s imperative!<br/>
	* [updated] BE filters<br/>
<br/>
`v0.564`<br/>
	* [fixed] issue where fresh-spawns are able to steal a not-owned vehicle in half a second<br/>
	* [added] if using the "lifted vehicles protection", there will also be a check for the owners, so stealing is not possible<br/>
<br/>
`v0.502`<br/>
	* [added] (optional) owner check for lifted vehicles (they also are protected now)<br/>
	* [added] new variable in clientside config.sqf (useVehLifted [true|false] recommended is true)<br/>
<br/>
`v0.444`<br/>
	* [fixed] if entered a safezone with a vehicle, you can get killed from a player standing outside of the zone<br/>
<br/>
`v0.443`<br/>
	* [fixed] issue with chopping trees after leaving a safezone (thx to ReDBaroN for noticing this bug)<br/>
	* [added] (optional) prohibit players to chop trees while in safezones (EPOCH_fnc_playerFired.sqf / check instructions in README.md)<br/>
	* [added] (optional) using chainsaws or sledgehammers in safezones will have no result (EPOCH_fnc_playerFired.sqf / check instructions in README.md)<br/>
	* [updated] BattlEye Filters<br/>
<br/>
`v0.436`<br/>
	* [added] option to unlock teleported vehicles (_unlockAfterVehTP	= 1; in serverside settings.h)<br/>
<br/>
`v0.432`<br/>
	* release<br/>


mpmissions/epoch.Mapname/SPKcode/ESVP/config.cpp
/*
	Title:				SPK_ESVP
	Author:				Sp4rkY [https://github.com/SPKcoding]
	Description:		Extended Safezones with Vehicle Protection for Arma 3 Epoch
	Version:			0.916
	Required:			Arma3 1.66+ / Epoch 0.4+
	File:				config.cpp (clientside config)
____________________________________________________________________________________________________________________________________________________________________________________________________________________*/


class CfgESVP {
	class safezones {															/* Define your safezones here. Can be as many as you want. (no comma after the last entry)											*/
		_defineSafezones[] = {													/* "ID": name of safezone (used for systemChat & accessGUI) | [x,y,z]: coordinates of safezone | rad: radius in meters (Number) 	*/
							{"WEST",	{6190.38,16842.4,0}, 250},				/* {"ID", {x,y,z}, rad},																											*/
							{"CENTER",	{13335.2,14508.6,0}, 250},				/* {"ID", {x,y,z}, rad},																											*/
							{"EAST",	{18461.1,14269.3,0}, 250}				/* {"ID", {x,y,z}, rad}																												*/
		};
	};
	class options {
		_useMarkers			= true;												/* Enable|Disable markers for the safezones at the map 																				*/
		_showNotes 			= true;												/* Show notifications (CfgNotifications) for entering/leaving the safezones															*/
		_showChatMsg 		= true;												/* Show systemChat messages for entering/leaving the safezones																		*/
		_useInfoMsg 		= true;												/* Show info messages while entering a safezone (true/false) 																		*/
		_useVehProt 		= true;												/* Use vehicle protection (if true, only owners can access their vehicles and it´s inventory in safezones)							*/
		_restrictDrv		= true;												/* Restrict driver seat to vehicle owner only [only if _useVehProt=true]															*/
		_useRestDrvMsg		= true;												/* Show systemChat message if joined driver seat [only if _restrictDrv=true]														*/
		_useVehLifted		= true;												/* Use check for lifted vehicles (similar to '_useVehProt' but for lifted/slingloaded vehicles) [only if _useVehProt=true]			*/
		_useSlingMsg		= true;												/* Show info messages while lifting prohibited vehicles in safezone [only if _useVehLifted=true] 									*/
		_useAccessGUI		= true;												/* Use "Access-Menu" for the driver (gain access to others to the vehicle while in safezone) [only if "_useVehProt=true"]			*/
		_usePlayerTP		= false;											/* Teleport players out of safezones at server restart to a random place in given min and max range									*/
		_pPlacePlrTP		= false;											/* Teleport players away from the parking place (if used in serverside settings) at server restart [only if "_usePlayerTP=true"]	*/
		_minRangePlrTP		= 250;												/* Minimum Range in meters the player is teleported to (should be minimum the same as the highest safezone radius) 					*/
		_maxRangePlrTP		= 350;												/* Maximum Range in meters the player is teleported to (have to be higher than "_minRangePlrTP")									*/
		_allowMelee			= false;											/* Enable|Disable melee weapons in safezones (hatchet, sledgehammer, chainsaw) [requires changes in EPOCH_fnc_playerFired.sqf]		*/
	};
	class markers {
		_mainColor			= "ColorGreen";										/* Color of the primary marker (Example: "ColorGreen", "ColorRed", "ColorBlue"...) 													*/
		_brush				= true;												/* Enable|Disable marker brush [if '_useMarkers=true'] 																				*/
		_brushStyle			= "Grid";											/* Set the style of the brush (examples: "Grid", "SolidBorder", "Cross") 															*/
		_useTextMarker		= true;												/* Enable|Disable text marker 																										*/
		_textColor			= "ColorRed";										/* Color of the text marker (Example: "ColorGreen", "ColorRed", "ColorBlue"...) [if '_useTextMarker=true']							*/
		_prefix				= false;											/* Enable|Disable prefix name of the markers 																						*/
		_prefixName			= "Safezone";										/* Define the prefix name here (markertext will be the prefix + ID of _defineSafezones: 'Safezone WEST', 'Safezone CENTER'...) 		*/
		_useUniqueName		= false;											/* If set to true, you will have a unique name for all markers (overrides '_prefix' & 'ID') [if '_useTextMarker=true']				*/
		_uniqueName			= "Safe area";										/* Define the markertext here [if '_useUniqueName=true'] 																			*/
		_textMrkType		= "mil_dot";										/* Type of the icon of the text marker (Example: "_mil_dot", "_mil_warning", "mil_join", "mil_circle"...) 							*/
	};
	class messages {
		class notes {															/* If (_showNotes=true) customise your notification messages below. (between the quotes "")											*/
			_enter[] 		= {"EnterSZ",{"You entered a Safezone!"}};			/* first: template-name (has to be the same as defined in CfgNotifications) | second: text you want to show (for entering the SZ)	*/
			_leave[] 		= {"LeaveSZ",{"You leave the Safezone!"}};			/* first: template-name (has to be the same as defined in CfgNotifications) | second: text you want to show (for leaving the SZ)	*/
		};
		class chat {
			_chatMsgEnter 	= "[INFO] YOU ENTERED SAFEZONE:";					/* systemChat message for entering a safezone between the quotes ""	(name of safezone is automatically added)						*/
			_chatMsgLeave 	= "[INFO] YOU LEAVED SAFEZONE:";					/* systemChat message for leaving a safezone between the quotes "" (name of safezone is automatically added)						*/
			_restDrvMsg		= "[INFO] DRIVER-SEAT IS RESTRICTED TO THE OWNER";	/* systemChat message while getting ejected from driver seat [only if _restrictDrv=true]											*/
		};
		class info {
			_infoMsg1 		= "safezone: access only to your own vehicle!";		/* Info message while entering a safezone (if _useVehProt=true) [Fires only if _useInfoMsg=true]									*/
			_infoMsg2		= "safezone: take care of your vehicle!";			/* Info message while entering a safezone (if _useVehProt=false) [Fires only if _useInfoMsg=true] 									*/
			_slingMsgOwnr	= "prohibited: you are not the owner!";				/* Info message while rope gets deleted if you are not the owner of the vehicle you want to lift [only if _useVehLifted=true] 		*/
			_slingMsgSteal	= "prohibited: vehicle wasn´t purchased by you!"; 	/* Info message while rope gets deleted if you have not purchased the vehicle you want to lift [only if _useVehLifted=true] 		*/
		};
	};
	class debug {
		_debugLog			= false;											/* Speaks for itself, right? (clientside RPT-log) [recommended: false, use it only if you have to]									*/
	};
};
#include "ctrl\gui.h"
